mediapipe/tasks/ios/genai/core/sources/LlmTaskRunner.swift

```diff
@@ -53,25 +52,99 @@ public final class LlmTaskRunner {
       }
     }
 
-    /// Throw an error if the response array is `NULL`.
+    /// Throw an error if response is invalid `NULL`.
+    guard let responseStrings = LlmTaskRunner.responseStrings(from: responseContext) else {
+      throw GenAiInferenceError.invalidResponse
+    }
+
+    return responseStrings
+  }
+
+  public func predict(
+    inputText: String, progress: @escaping (_ partialResult: [String]?, _ error: Error?) -> Void,
+    completion: @escaping (() -> Void)
+  ) {
+
+    /// `strdup(inputText)` prevents input text from being deallocated as long as callbacks are
+    /// being invoked. `CallbackInfo` takes care of freeing the memory of `inputText` when it is
+    /// deallocated.
+    let callbackInfo = CallbackInfo(
+      inputText: strdup(inputText), progress: progress, completion: completion)
+    let callbackContext = UnsafeMutableRawPointer(Unmanaged.passRetained(callbackInfo).toOpaque())
+
+    LlmInferenceEngine_Session_PredictAsync(cLlmSession, callbackContext, callbackInfo.inputText) {
+      context, responseContext in
+      guard let cContext = context else {
+        return
+      }
+
+      /// `takeRetainedValue()` decrements the reference count incremented by `passRetained()`. Only
+      /// take a retained value if the LLM has finished generating responses to prevent the context
+      /// from being deallocated in between response generation.
+      let cCallbackInfo =
+        responseContext.done
+        ? Unmanaged<CallbackInfo>.fromOpaque(cContext).takeRetainedValue()
+        : Unmanaged<CallbackInfo>.fromOpaque(cContext).takeUnretainedValue()
+
+      if let responseStrings = LlmTaskRunner.responseStrings(from: responseContext) {
+        cCallbackInfo.progress(responseStrings, nil)
+      } else {
+        cCallbackInfo.progress(nil, GenAiInferenceError.invalidResponse)
+      }
+
+      /// Call completion callback if LLM has generated its last response.
+      if responseContext.done {
+        cCallbackInfo.completion()
+      }
+    }
+  }
+
+  deinit {
+    LlmInferenceEngine_Session_Delete(cLlmSession)
+  }
+}
+
+extension LlmTaskRunner {
+  /// A wrapper class for whose object will be used as the C++ callback context.
+  /// The progress and completion callbacks cannot be invoked without a context.
+  class CallbackInfo {
+    typealias ProgressCallback = (_ partialResult: [String]?, _ error: Error?) -> Void
+    typealias CompletionCallback = () -> Void
+
+    let inputText: UnsafeMutablePointer<CChar>?
```

Comment:
This raised several questions to be solved in the (near?) future: 
* shall we let the caller know that CallbackInfo takes ownership of inputText?
* or: do we even need to make inputText a member of callback info? it's not immediately obvious why we need to retain the value after the initial call

Reply:
if the C APIs are not copying the String before delegating to the background thread, it will get deallocated since its lifetime is only for the duration of this method. The method is asynchronous and returns before the response generation callbacks are invoked. So adding them to the context was my solution for this. We can also make this an instance variable in TaskRunner since we don't allow simultaneous calls. But if in future, we change this behaviour, we won't be able to keep track of inputText's from all calls.

I am okay removing this provided the C APIs copy the string. I didn't see that in the C code.

Reply:
Also the CallbackInfo does not take ownership of the inputText. We allocate a new copy with strdup(). CallbackInfo takes ownership of this copy and deallocates it in deinit